legacy/firmware/crypto.c

```diff
@@ -975,3 +978,112 @@ bool multisig_uses_single_path(const MultisigRedeemScriptType *multisig) {
     return true;
   }
 }
+
+// descriptor checksum manually translated from
+// https://github.com/bitcoin-core/HWI/blob/master/hwilib/descriptor.py
+
+static const char INPUT_CHARSET[] =
+    "0123456789()[],'/*abcdefgh@:$%{}"
+    "IJKLMNOPQRSTUVWXYZ&+-.;<=>?!^_|~"
+    "ijklmnopqrstuvwxyzABCDEFGH`#\"\\ ";
+static const char CHECKSUM_CHARSET[] = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+static const size_t DESCRIPTOR_CHECKSUM_LEN = 8;
+
+static uint64_t descriptor_polymod(uint64_t c, uint8_t val) {
+  uint64_t c0 = c >> 35;
+  c = ((c & 0x7FFFFFFFF) << 5) ^ val;
+  if (c0 & 1) {
+    c ^= 0xF5DEE51989;
+  }
+  if (c0 & 2) {
+    c ^= 0xA9FDCA3312;
+  }
+  if (c0 & 4) {
+    c ^= 0x1BAB10E32D;
+  }
+  if (c0 & 8) {
+    c ^= 0x3706B1677A;
+  }
+  if (c0 & 16) {
+    c ^= 0x644D626FFD;
+  }
+  return c;
+}
+
+static bool descriptor_checksum(const char *descriptor, size_t descriptor_len,
+                                char dest[DESCRIPTOR_CHECKSUM_LEN]) {
+  uint64_t c = 1;
+  uint8_t cls = 0;
+  uint8_t clscount = 0;
+
+  for (size_t i = 0; i < descriptor_len; i++) {
+    char *pos_ptr = strchr(INPUT_CHARSET, descriptor[i]);
+    if (!pos_ptr || !*pos_ptr) {
+      return false;
+    }
+    size_t pos = pos_ptr - INPUT_CHARSET;
+    c = descriptor_polymod(c, pos & 31);
+    cls = cls * 3 + (pos >> 5);
+    clscount += 1;
+    if (clscount == 3) {
+      c = descriptor_polymod(c, cls);
+      cls = 0;
+      clscount = 0;
+    }
+  }
+  if (clscount > 0) {
+    c = descriptor_polymod(c, cls);
+  }
+  for (size_t j = 0; j < 8; j++) {
+    c = descriptor_polymod(c, 0);
+  }
+  c ^= 1;
+  for (size_t j = 0; j < DESCRIPTOR_CHECKSUM_LEN; j++) {
+    size_t pos = (c >> (5 * (7 - j))) & 31;
+    if (pos >= sizeof(CHECKSUM_CHARSET)) {
+      return false;
+    }
+    dest[j] = CHECKSUM_CHARSET[pos];
+  }
+  return true;
+}
+
+bool descriptor_format(InputScriptType script_type, uint32_t root_fingerprint,
+                       const uint32_t address_n[], size_t address_n_count,
+                       const char *xpub, char *dest, size_t dest_size) {
+  char *type = NULL;
+  char *terminator = "";
+  if (script_type == InputScriptType_SPENDADDRESS) {
+    type = "pkh";
+  } else if (script_type == InputScriptType_SPENDP2SHWITNESS) {
+    type = "sh(wpkh";
+    terminator = ")";
+  } else if (script_type == InputScriptType_SPENDWITNESS) {
+    type = "wpkh";
+  } else if (script_type == InputScriptType_SPENDTAPROOT) {
+    type = "tr";
+  } else {
+    return false;
```

Comment:
Would it be possible to add a test vector to cover this case?

Reply:
sure: efa7945